specs/HostResourceMapping.md

```diff
@@ -0,0 +1,108 @@
+# Background
+We have received feedbacks that there is a need to access local files in WebView2. In this document we describe the APIs for this scenario. We'd appreciate your feedback.
+
+
+# Description
+To access local files in WebView2, define a virutal host name to local folder mapping using the `AddVirtualHostNameToFolderMapping` API,
+and then refer to local files under the folder with a normal http/https url with the virtual host name.
+
+# Examples
+Suppose the app has index.html file next to its exe file. The following code snippet demonstrates how to define a mapping
+for `app-file.invalid` host name so that it can be accessed from pages using normal http/https url like https://app-file.invalid/index.html.
+
+## Win32 C++
+```cpp
+webview2->AddVirtualHostNameToFolderMapping(
+    L"app-file.invalid", L".", COREWEBVIEW2_HOST_RESOURCE_ACCESS_KIND_DENY_CORS);
+webview2->Navigate("https://app-file.invalid/index.html");
+```
+
+## .Net and WinRT
+```c#
+webView.CoreWebView2.AddVirtualHostNameToFolderMapping(
+    "app-file.invalid", ".", CoreWebView2HostResourceAccessKind.DenyCors);
+webView.Navigate("https://app-file.invalid/index.html");
+```
+
+# Remarks
+Choose a virtual host name that will not be used by real sites.  
+If you own a domain (like example.com), an option is to use a sub domain reserved for the app (like my-app.example.com).  
+If you don't own a domain, [RFC 6761](https://tools.ietf.org/html/rfc6761) has reserved several special-use domain names that would
+not be used by real sites, like .example, .test, and .invalid.
```

Comment:
Personally I think it's a bad idea to encourage anyone (including in samples) to use .invalid in any code anywhere unless you specifically WANT it to fail, which I don't think is the case here.  I.e. if you were to pass the name to HostName constructor and the constructor didn't accept it, your code should still be correct.  As such I would claim line 23 of this PR is a bug.

Reply:
We were basing this off of https://tools.ietf.org/html/rfc6761#section-6.4  which says applications may recognize `.invalid` domains as special. But it sounds like you're saying the way in which applications should recognize these as special is to make them fail in some manner. Additionally it does look strange to use `.invalid` in code.

Do you have another recommendation instead of using `.invalid`? We also are considering reserving a domain name that end devs could use for this purpose.

Reply:
`.example` and `.test` are probably better choices than `.invalid`. Even if it works, it sure smells weird. (Looks like SIP uses `anonymous.invalid` as an intentionally-invalid domain name to represent an anonymous caller.)

Reply:
The RFC says applications should not treat `.test` special and should be resolved like other hostnames. Same for `.example`. I think the only good answer is to reserve a domain for this purpose like they do in the similar Android API. We're looking into that now. It could be a subdomain off an existing Microsoft domain.

Reply:
`.example` and `.test` are confusing though, they look test code rather than product. And the RFC makes it sound like this is the intended use for `invalid.`:

> Application software MAY recognize "invalid" names as special or MAY pass them to name resolution APIs as they would for other domain names.

Reply:
Would be good to add that URL (https://tools.ietf.org/html/rfc6761#section-6.4) to the doc to explain.  I agree that .invalid is better than .example or .test if we have to use a hostname at all.   Using a separate domain might be cleaner (like we did with ipv6-literal.net) but you do have to worry about DNS queries actually being sent for it by some software.

Reply:
Updated. Reserved a special domain name for apps to use.

Reply:
A note, there are security implications of using reserved appassets.webview2.microsoft.com host name as it is considered as a sub domain of the real microsoft.com. We are going to switch back to suggesting app uses .example/.test/.invalid host names.